### DISCUSSION TITLE
Make libp2p::request_response::RequestId instantiable?

### DISCUSSION THREAD

**Discussion by matthias-wright** [libp2p/rust-libp2p#3712] [Q&A] [answered]
Hey all,

Thanks for the awesome work you do!
I wanted to ask if there are any objections against making [libp2p::request_response::RequestId](https://docs.rs/libp2p/latest/libp2p/request_response/struct.RequestId.html) instantiable?
If the field should remain private, what about adding a method like [PeerId::random()](https://docs.rs/libp2p/latest/libp2p/struct.PeerId.html#method.random)?

It would be helpful to create instances of `RequestId` to write units tests.
I can create a PR for this in case this is something you would consider.

Cheers!

**Comment by mxinden** [comment, marked as answer]
Hi Matthias,

> It would be helpful to create instances of RequestId to write units tests.

Can you expand on why you need to instantiate a `RequestId` in a unit test? I am assuming a `libp2p-request-response` user unit test?

In general, I would like to keep the interface of `libp2p-*` crates as small as possible as that allows us to package more changes as non-breaking, and thus easy for all of you to upgrade to.

**Reply by matthias-wright** [reply]
Hi @mxinden, thanks the the response!
> I am assuming a libp2p-request-response user unit test?

Yes exactly, it's for `libp2p-request-response` user unit tests.

> In general, I would like to keep the interface of libp2p-* crates as small as possible as that allows us to package more changes as non-breaking, and thus easy for all of you to upgrade to.

Yeah that makes sense. I found a workaround to create the tests, it would look a bit cleaner if I could instantiate a `RequestId` but the maintainability of the crates is definitely more important.

**Reply by thomaseizinger** [reply]
There is also `libp2p-swarm-test` which makes writing tests for behaviours through a `Swarm` a lot easier. Maybe that helps.
